src/shutdown.rs

```diff
@@ -281,29 +282,24 @@ impl SourceShutdownCoordinator {
         name: String,
         deadline: Instant,
     ) -> impl Future<Item = bool, Error = ()> {
-        let success = shutdown_complete_tripwire.map(move |_| true);
-
-        let timeout = timer::Delay::new(deadline)
-            .map(move |_| {
-                error!(
-                    "Source '{}' failed to shutdown before deadline. Forcing shutdown.",
-                    name,
-                );
-                false
-            })
-            .map_err(|err| panic!("Timer error: {:?}", err));
-
-        let union = success.select(timeout);
-        union
-            .map(|(success, _)| {
-                if success {
+        let done = async move {
+            select! {
+                _ = shutdown_complete_tripwire.compat() => {
                     shutdown_force_trigger.disable();
-                } else {
+                    true
+                }
+                _ = time::delay_until(deadline.into()) => {
+                    error!(
+                        "Source '{}' failed to shutdown before deadline. Forcing shutdown.",
+                        name,
+                    );
                     shutdown_force_trigger.cancel();
+                    false
                 }
-                success
-            })
-            .map_err(|_| ())
+            }
+        };
+
+        Box::pin(done.map(Ok)).compat()
```

Comment:
For method chaining `fut.boxed()` can be used `Box::pin(fut)` :slightly_smiling_face: 
https://docs.rs/futures/0.3.5/futures/future/trait.FutureExt.html#method.boxed